### ISSUE TITLE
Scopes aren't working with AR::Store & Postgres Jsonb

### ISSUE THREAD

**Issue by radubogdan** [rails/rails#26650]
### Steps to reproduce
1. Have a jsonb column
2. Use AR::Store
3. Create a scope for a "json attribute"
### Expected behavior

It should return the records matched by that scope.
### Actual behavior

It returns []

My guess is that because it's serialized in the db, normal queries doesn't work.
### Example code:

Check [this](https://github.com/radubogdan/rails-5-ar-store/tree/master/app/models) dummy app. The only difference between `Car` and `BadCar` model is that `BadCar` is using `AR::Store`. I'd have expected same results.

**Because it's better to see some code:**

```
# BadCar is the one that is using AR::Store

BadCar.create(down_payment: "1000")
   (0.1ms)  BEGIN
  SQL (2.4ms)  INSERT INTO "bad_cars" ("financing_offer") VALUES ($1) RETURNING "id"  [["financing_offer", "\"{\\\"down_payment\\\":\\\"1000\\\"}\""]]
   (7.1ms)  COMMIT
=> #<BadCar id: 1, financing_offer: {"down_payment"=>"1000"}>
```

```
BadCar.financing_offer_eq(:down_payment, "1000")
  BadCar Load (0.2ms)  SELECT "bad_cars".* FROM "bad_cars" WHERE (bad_cars.financing_offer->>'down_payment' = '1000')
=> #<ActiveRecord::Relation []>
```

```
# Without AR::Store, just jsonb

Car.create(financing_offer: { down_payment: "1000"})
   (0.1ms)  BEGIN
  SQL (0.4ms)  INSERT INTO "cars" ("financing_offer", "created_at", "updated_at") VALUES ($1, $2, $3) RETURNING "id"  [["financing_offer", "{\"down_payment\":\"1000\"}"], ["created_at", 2016-09-28 12:49:02 UTC], ["updated_at", 2016-09-28 12:49:02 UTC]]
   (1.3ms)  COMMIT
=> #<Car id: 1, financing_offer: {"down_payment"=>"1000"}, created_at: "2016-09-28 12:49:02", updated_at: "2016-09-28 12:49:02">
```

```
Car.financing_offer_eq(:down_payment, "1000")
  Car Load (0.6ms)  SELECT "cars".* FROM "cars" WHERE (cars.financing_offer->>'down_payment' = '1000')
=> #<ActiveRecord::Relation [#<Car id: 1, financing_offer: {"down_payment"=>"1000"}, created_at: "2016-09-28 12:49:02", updated_at: "2016-09-28 12:49:02">]>
```
### System configuration

**Rails version**: 5

**Ruby version**: 2.3.1

**Comment by maclover7** [comment]
Can reproduce on master with [this reproduction script](https://gist.github.com/maclover7/faa3ce80e52c0f89d4163869d323c008).

**Comment by uaihebert** [comment]
I do not know if the problem was clear for you guys, but I was looking at the data saved in the database and there is a difference:

![image](https://cloud.githubusercontent.com/assets/2014377/18972803/01c1d546-8671-11e6-8179-62c7428b65d3.png)

Reading the docs I could see that the `store` is not saved to be accessed as a regular field but as hash. Looks like the problem is in how the field is queried?

None of the queries bellow could find the column:
```
select * from bad_cars where financing_offer-->'"down_payment' = '1000';
select * from bad_cars where financing_offer-->'"down_payment"' = '1000';
select * from bad_cars where financing_offer-->'\"down_payment\"' = '1000';
```

A terrible way for doing query would be like:
```
select * from bad_cars where financing_offer::text like '%1000%';
```

I am thinking that it could be considered a that store are not mean to be accessible as JSON.

PS.: I found this old article that says that the only way to query a `store` field is by `like`: http://samuelmullen.com/2012/04/activerecord-base-store-pros-cons-usage/

**Comment by radubogdan** [comment]
Thank you @maclover7 for the script and @uaihebert for jumping in and post your insights about this. I've also noticed the difference in the db and tried almost every query I could've think of.

The article you posted is nice and looks like that's the only way to query a store field. However it was written in 2012 and I think more functionality was added on top and there should be a fix to query it the normal way.

**Comment by kamipo** [comment]
Seems `store_accessor :financing_offer, :down_payment` works.
https://github.com/rails/rails/blob/master/activerecord/test/cases/adapters/postgresql/json_test.rb#L10

**Comment by uaihebert** [comment]
The value is stored without a problem and recovered with the model without a problem. The problem will happen if a query is fired to this attribute. But I do not think that this is a real problem since the attribute is not saved as a valid json in the database.

**Comment by radubogdan** [comment]
@uaihebert You're right. Any ideas why the attribute is not saved as a valid json?

**Comment by NielsKSchjoedt** [comment]
This one is slightly related to this: https://github.com/rails/rails/issues/25594

**Comment by NielsKSchjoedt** [comment]
I investigated the issue further, and what seems to be the underlying issue is that rails/pg already serialises/deserialize the attributes, when you use JSONB type. Now this means that the `store :attribute, coder: JSON` directive should be OMITTED in rails 5! Else the data is serialises/deserialised twice.

**Comment by kamipo** [comment]
Addressed by #29520.
